app/src/main/java/com/globant/calculatoronboardingkotlin/mvp/presenter/CalculatorPresenter.kt

```diff
@@ -15,20 +23,98 @@ class CalculatorPresenter(
+        if ((model.firstNumber.isNotEmpty()) && (model.secondNumber.isNotEmpty()) &&
+            (model.firstNumber != ZERO_DOT) && (model.secondNumber != ZERO_DOT)
+        ) {
+            view.showResult(calculateResult())
+            model.clear()
+        } else if (model.operator.isEmpty() && (model.firstNumber != ZERO_DOT)) {
+            view.showResult(model.firstNumber)
+            model.clear()
+        } else {
+            view.showError(ErrorMessages.INVALID_OPERATION)
+            onClearButtonPressed()
+        }
     }
 
     override fun onNumberButtonPressed(number: String) {
-        model.firstNumber = number
-        view.showInputPressed(model.firstNumber)
+        if (model.operator == EMPTY_STRING) {
+            if (model.secondNumber.isEmpty()) {
+                model.firstNumber = "${model.firstNumber}$number"
+                view.showInputPressed(model.firstNumber)
+            }
+        } else {
+            model.secondNumber = "${model.secondNumber}$number"
+            view.showInputPressed(model.secondNumber)
+        }
     }
+
+    override fun onDotButtonPressed(dot: String) {
+        if ((model.secondNumber.isEmpty()) && (model.operator.isEmpty())) {
+            if (model.firstNumber.isEmpty()) {
+                model.firstNumber = "${NUMBER_ZERO}$dot"
+                view.showInputPressed(model.firstNumber)
+            } else if (!model.firstNumber.contains(DOT)) {
+                model.firstNumber = "${model.firstNumber}$dot"
+                view.showInputPressed(model.firstNumber)
+            } else {
+                view.showError(ErrorMessages.TOO_MANY_DOTS)
+            }
+        } else if (model.operator.isNotEmpty()) {
+            if (model.secondNumber.isEmpty()) {
+                model.secondNumber = "${NUMBER_ZERO}$dot"
+                view.showInputPressed(model.secondNumber)
+            } else if (!model.secondNumber.contains(DOT)) {
+                model.secondNumber = "${model.secondNumber}$dot"
+                view.showInputPressed(model.secondNumber)
+            } else {
+                view.showError(ErrorMessages.TOO_MANY_DOTS)
+            }
+        }
+    }
+
+    private fun calculateResult(): String {
+        val firstNumber = model.firstNumber.toDouble()
+        val secondNumber = model.secondNumber.toDouble()
+        var result: Double = ZERO_DOUBLE
+        when (model.operator) {
+            PLUS -> result = firstNumber + secondNumber
+            MINUS -> result = firstNumber - secondNumber
+            MULTIPLY -> result = firstNumber * secondNumber
+            DIVIDE -> if (secondNumber == ZERO_DOUBLE) {
+                onClearButtonPressed()
+                view.showError(ErrorMessages.DIVIDE_BY_ZERO)
+            } else {
+                result = firstNumber / secondNumber
+            }
+        }
+        return result.toString()
+    }
+}
+
+enum class ErrorMessages(val message: String) {
+    DIVIDE_BY_ZERO("Tried to divide by zero"),
```

Comment:
Can you put that strings in Utils or other side?

Reply:
done, thanks